client/CVideoHandler.cpp

```diff
@@ -636,26 +648,92 @@ bool CVideoPlayer::playVideo(int x, int y, bool stopOnKey, bool overlay)
 
 	auto lastTimePoint = boost::chrono::steady_clock::now();
 
+	// determine a resolution that has the 800:600 aspect ratio and fits inside the selected VCMI resolution
+	float resX = settings["video"]["resolution"]["width"].Float(); // Float, since we do some floating point calculations
+	float resY = settings["video"]["resolution"]["height"].Float();
+	int originalH3ResX = 800;
+	int originalH3ResY = 600;
+	float aspectRatio_OH3 = (float) originalH3ResX/originalH3ResY;
+	float resYRatioVCMI_OH3 = resY/originalH3ResY; // 0..1
+	int correctedResX = originalH3ResX*resYRatioVCMI_OH3;
+	int correctedResY = originalH3ResY*resYRatioVCMI_OH3;
+
+	SDL_Surface *image;
+	SDL_Texture *backgroundTexture = nullptr;
+	SDL_Rect backgroundRect;
+	SDL_Rect videoRect;
+
+	int H3INTROResX = 640;
+	int H3INTROResY = 360;
+
+	if(showFrame)
+	{
+		ImagePath imageToOpen = ImagePath::builtin("INTRORIM.bmp");
+		ImagePath iname;
+
+		if (CResourceHandler::get()->existsResource(imageToOpen))
+			iname = imageToOpen;
+		else
+			iname = imageToOpen.addPrefix("DATA/");
+
+		if (!CResourceHandler::get()->existsResource(iname))
+		{
+			logGlobal->error("Error: image %s was not found", iname.getName());
+			return false;
+		}
+
+		image = BitmapHandler::loadBitmap(iname);
+
+		if (!image) {
+			logGlobal->error("Error: failed to load image at %s: %s\n", iname.getName(), SDL_GetError());
+			return false;
+		}
+
+		int offset = 0;
+		backgroundRect = CSDL_Ext::toSDL(Rect(offset, 0, correctedResX, correctedResY));
+		backgroundTexture = SDL_CreateTextureFromSurface(mainRenderer, image);
+		videoRect = CSDL_Ext::toSDL(Rect(offset+pos.x*correctedResX/originalH3ResX, pos.y*correctedResY/originalH3ResY, H3INTROResX*correctedResX/originalH3ResX, H3INTROResY*correctedResY/originalH3ResY));
+	} else {
+		videoRect = CSDL_Ext::toSDL(Rect(pos.x, pos.y, correctedResX, correctedResY));
+	}
+
 	while(nextFrame())
 	{
+		
 		if(stopOnKey)
 		{
 			GH.input().fetchEvents();
 			if(GH.input().ignoreEventsUntilInput())
 				return false;
 		}
 
-		SDL_Rect rect = CSDL_Ext::toSDL(pos);
-
 		if(overlay)
 		{
+			SDL_Rect rect = CSDL_Ext::toSDL(pos);
 			SDL_RenderFillRect(mainRenderer, &rect);
 		}
 		else
 		{
 			SDL_RenderClear(mainRenderer);
 		}
-		SDL_RenderCopy(mainRenderer, texture, nullptr, &rect);
+
+		correctedResX = aspectRatio_OH3*resY;
+		correctedResY = resY;
+
+		int offsetX = (resX-correctedResX)/2;
+		int offsetY = (resY-correctedResY)/2;
+		if(showFrame)
+		{
+			backgroundRect = CSDL_Ext::toSDL(Rect(offsetX, offsetY, correctedResX, correctedResY));
+			SDL_RenderCopy(mainRenderer, backgroundTexture, nullptr, &backgroundRect);
+			videoRect = CSDL_Ext::toSDL(Rect(	offsetX+pos.x*correctedResX/originalH3ResX,
```

Comment:
Looks like this with a tab size of 4:

![image](https://github.com/vcmi/vcmi/assets/3226457/dad363fd-3679-43e5-810a-2f9503d459ff)